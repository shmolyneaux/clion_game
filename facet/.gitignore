target
.docs-cache
.aider*
*~lock~
/templates/output.rs
tarpaulin-report.html
.env
*long-type-*.txt
.ignore.rs
**/.claude/settings.local.json
**/CLAUDE.local.md
*.sw[op]
