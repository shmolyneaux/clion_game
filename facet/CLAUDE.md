### GitHub Repository

The main GitHub repository for Facet is at https://github.com/facet-rs/facet

### READMEs

Don't edit any `README.md` files directly â€” edit `README.md.in` and run `just
codegen` to regenerate the READMEs.

Run `just` to make sure that tests pass.

For doc tests, you need to do `just codegen && just doc-tests`. Remember to fix
them by editing the corresponding `README.md.in`, not `README.md`.

All crates have their own readme template, except for the `facet/` crate, which
has it in the top-level `README.md.in`

### Testing

Always use `cargo nextest run` instead of `cargo test` to run tests. Nextest provides better isolation between tests and avoids issues with shared test environments.

For example:
- Run a specific test: `cargo nextest run test_name`
- Run tests in a specific module: `cargo nextest run module_name`
- Run tests with debug output: `cargo nextest run --no-capture test_name`

When debugging serialization/deserialization issues, you can enable more verbose logging with the `facet-reflect/log` feature flag:
```bash
cargo nextest run -F facet-reflect/log my_test
```

To run tests under Miri (the memory safety checker), use the nightly toolchain:
```bash
cargo +nightly miri nextest run -p facet-reflect test_name
```

### Pre-commit Hooks

When committing changes, facet-dev will run to check for code generation and formatting changes.
It will automatically apply any necessary fixes without requiring user input.

### Git Force Push Safety

Always use `git push --force-with-lease` instead of `git push --force` when force-pushing changes.
The `--force-with-lease` option provides a safety check that prevents overwriting others' work that
you haven't seen yet, making it safer for collaborative development.

### Tuple implementations

The file `tuples_impls.rs` in facet-core is generated from `gen_tuples_impls.rs`
in the `facet-codegen` crate. If you see any errors in it, do not correct them,
simply make a note of it and I will take care of it.

### Dependencies

crates like `facet-yaml`, `facet-json`, only have have a _dev_ dependency on
`facet`. For regular dependencies, they only have `facet-core`, `facet-reflect`.
This is to keep `facet-macros` out of them.

### Testing Derive Macros

Tests that exercise the `#[derive(Facet)]` macro cannot live in `facet-core`
because it does not depend on `facet-macros`. Such tests should either be
snapshot tests in `facet-macros-emit` or integration tests in the main `facet`
crate, which brings all the necessary components together.

### Def and Type enums

In the facet system, there are two separate enum types that describe types:

- `Type`: Represents the Rust type system classification. This includes:
  - `Type::User`: User-defined types like structs, enums, and unions
  - `Type::Sequence`: Sequence types like tuples, arrays, etc.
  - `Type::Primitive`: Built-in primitive types
  - `Type::Pointer`: Reference and pointer types

- `Def`: Represents common, well-known data structures for interacting with values:
  - `Def::Map`: Dictionary or map-like structures
  - `Def::List`: Ordered list or sequence of homogeneous values
  - `Def::Array`: Fixed-size homogeneous arrays
  - `Def::Option`: Optional values
  - `Def::Scalar`: Simple scalar values
  - `Def::Undefined`: Used when no specific `Def` applies; in this case, check `Type`

When working with type information:
1. First check `Def` for common collection types like maps, lists, etc.
2. For user-defined types, use `Type::User` and check for `UserType::Struct`, `UserType::Enum`, etc.
3. For tuples, use `Type::Sequence(SequenceType::Tuple)`.

This design lets facet handle both generic data structures (`Def`) and Rust's specific type system (`Type`).

### Type Conversion in Deserialization

When implementing deserialization in format-specific modules (YAML, JSON, etc.):

- The `facet_deserialize` infrastructure handles automatic conversions for certain scalar types
- For types with specific `ScalarAffinity` (Time, UUID, Path, etc.), string values are automatically converted
- Use `partial.set(string_value)` and the system will handle parsing if the target type supports it
- This keeps format-specific code simple and avoids duplication of parsing logic

Use `ScalarAffinity` patterns to detect and handle related types:

```rust
// For scalar_def with time affinity (like OffsetDateTime)
if matches!(scalar_def.affinity, ScalarAffinity::Time(_)) {
    // Simply set the string value, the automatic conversion will handle parsing
    let s = value.as_str().unwrap_or_default().to_string();
    wip.set(s).map_err(|e| AnyErr(e.to_string()))?;
}
```

Other common scalar affinities to handle:
- `ScalarAffinity::Path(_)` - for Path types
- `ScalarAffinity::UUID(_)` - for UUID types 
- `ScalarAffinity::ULID(_)` - for ULID types

### Code Comments

Avoid adding comments that merely restate what the code is doing or that reference the development process (e.g., "BUG:", "TODO:" unless they're meant to stay). Comments should add value by explaining complex logic or design decisions, not narrate the obvious or temporary state of the code.