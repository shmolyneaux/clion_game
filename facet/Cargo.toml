[workspace]
members = [
    # core
    "facet-core",
    "facet",

    # derive macro
    "facet-macros-parse",
    "facet-macros-emit",
    "facet-macros",

    # reflection
    "facet-reflect",

    # dev helpers
    "facet-dev",

    # # test helpers
    "facet-testhelpers",

    # # formats / ecosystem
    "facet-args",
    "facet-csv",
    "facet-json",
    "facet-msgpack",
    "facet-serialize",
    "facet-pretty",
    "facet-toml",
    "facet-kdl",
    "facet-urlencoded",
    "facet-xdr",
    "facet-yaml",
    "facet-jsonschema",
    "facet-deserialize",
    "facet-bench",
    "facet-testhelpers-macros",
]
exclude = ["outside-workspace"]
resolver = "3"

[workspace.package]
edition = "2024"
rust-version = "1.87.0"
license = "MIT OR Apache-2.0"
repository = "https://github.com/facet-rs/facet"
