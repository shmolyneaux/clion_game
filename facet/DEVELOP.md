# facet development guide

## Website

The project's website is [facet.rs](https://facet.rs). The website source files
can be found in the `docs/` directory.

Only @fasterthanlime can deploy the website but anyone can run it locally
because [home](https://home.bearcove.cloud) is now open-source.

## Collaboration and Contribution Guidelines

Try to submit changes as pull requests (PRs) for review and feedback, even if
you're part of the organization. We all champion different things: @fasterthanlime
facet-json, @tversteeg face-toml, @Veykril language stuff, @epage has good
advice when it comes to crate design and no_std — and other stuff!

## Pull Request Best Practices

Prefer smaller, incremental PRs over large, monolithic ones. This helps avoid
stagnation and makes review easier, even though initial setup PRs may be large.

## Staying Up to Date

Expect some churn as APIs evolve. Keep up with changes in core libraries (like
facet-reflect and facet-core) as needed. Coordination during rapid development
is key.

## Version Control & Checks

You’re welcome to use alternative version control tools (like jujutsu/jj), but
always run checks such as `just precommit`, `just prepush`, or `just ci` before
merging to avoid CI failures.

## Regenerating Documentation

Use `just gen` to regenerate the `README.md` whenever documentation needs to be
updated — that's normally part of the precommit hook.

## Precommit/prepush hook

Sometimes the hook just won't pass, and in that case you can just pass
`--no-verify` to either `git commit` or `git push`. Nobody'll get mad at you
except for the CI pipeline.

Those hooks are only here to save you from back-and-forths with CI! They should
serve you, they're a sign, not a cop, etc.

## Shipping

Only @fasterthanlime has publish rights for the crates.

They run `just ship` locally — which uses [release-plz](https://release-plz.ieni.dev).

## Running tests

Do yourself a favor, run tests with [cargo-nextest](https://nexte.st) — using
`cargo test` is _not officially supported_.

Make sure to check the platform-specific notes:

  * [for macOS](https://nexte.st/docs/installation/macos/)
  * [for Windows](https://nexte.st/docs/installation/windows/)

As of Apr 22, 2025, the 349 tests run in .742s on a MacBook Pro M4.

## Rust nightly / MSRV

facet does not use Rust nightly, on purpose. It is "the best of stable". However,
the MSRV will likely bump with every new Rust stable version for the foreseeable
future.
