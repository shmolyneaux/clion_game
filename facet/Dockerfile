# Allow selecting base image at build time
ARG BASE_IMAGE=rust:1.87-slim-bookworm
FROM ${BASE_IMAGE} AS builder

ARG RUSTUP_TOOLCHAIN
ENV RUSTUP_TOOLCHAIN=${RUSTUP_TOOLCHAIN}

ENV CI=true

# Accept additional Rust components as a build argument
ARG ADDITIONAL_RUST_COMPONENTS=""
ENV ADDITIONAL_RUST_COMPONENTS=${ADDITIONAL_RUST_COMPONENTS}

# Install Rust components
RUN rustup component add rust-src clippy rustfmt llvm-tools-preview ${ADDITIONAL_RUST_COMPONENTS}

# Set the default toolchain based on build arg and configure rust components
RUN apt-get update && apt-get install -y curl libssl-dev git-core pkg-config && \
    # Install cargo-binstall using curl
    curl -L --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/cargo-bins/cargo-binstall/main/install-from-binstall-release.sh | bash && \
    # Clean up - remove curl and apt cache
    apt-get purge -y curl && \
    apt-get autoremove -y && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Install tools using cargo-binstall
RUN cargo binstall -y just cargo-nextest cargo-llvm-cov

# Add the thumbv8m.main-none-eabihf Rust target
RUN rustup target add thumbv8m.main-none-eabihf

# Set environment variables
ENV CARGO_INCREMENTAL=0
ENV CLICOLOR=1
ENV FORCE_COLOR=1
ENV CLICOLOR_FORCE=1

# Create a work directory
WORKDIR /app

CMD ["bash"]

LABEL org.opencontainers.image.source="https://github.com/facet-rs/facet"
