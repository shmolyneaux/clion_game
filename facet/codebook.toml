dictionaries = []
words = [
    "def",
    "facet",
    "guard",
    "vtable",
]
flag_words = []
ignore_paths = []
ignore_patterns = []
use_global = true
