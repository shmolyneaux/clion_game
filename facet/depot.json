{"id":"8531f2vcgf"}
