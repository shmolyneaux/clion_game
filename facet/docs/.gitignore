/node_modules
/stats.html
/.home
/dist
.envrc
.env
