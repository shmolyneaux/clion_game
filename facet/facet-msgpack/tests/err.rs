use eyre::Result;
use facet::Facet;
use facet_msgpack::from_slice;

#[derive(Facet, Debug)]
struct FooBar {
    foo: u64,
    bar: String,
}

#[test]
fn bad_hex_1() -> Result<()> {
    facet_testhelpers::setup();

    let data = [
        0x82, // Fixmap with 2 elements
        0xa3, // Fixstr with length 3
        0x66, 0x6f, 0x6f, // "foo"
        0xce, // uint32 (correct prefix according to MessagePack spec)
        0x00, 0x00, 0x00, 0x2a, // 42
    ];
    let err = from_slice::<FooBar>(&data).unwrap_err();
    insta::assert_snapshot!(err);
    Ok(())
}

#[test]
fn bad_hex_2() -> Result<()> {
    facet_testhelpers::setup();

    let data = [
        0x82, // Fixmap with 2 elements
        0xa3, // Fixstr with length 3
        0x66, 0x6f, 0x6f, // "foo"
        0xce, // uint32 (correct prefix according to MessagePack spec)
        0x00, 0x00, 0x00, 0x2a, // 42
        0xa3, // Fixstr with length 3
        0x62, 0x61, 0x72, // "bar"
        0xce, // uint32 (correct prefix according to MessagePack spec)
        0x00, 0x00, 0x00, 0x2a, // 42
    ];
    let err = from_slice::<FooBar>(&data).unwrap_err();
    insta::assert_snapshot!(err);
    Ok(())
}
