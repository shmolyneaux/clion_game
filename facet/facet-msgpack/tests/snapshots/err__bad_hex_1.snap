---
source: facet-msgpack/tests/err.rs
expression: err
---
Insufficient data to decode
